### ISSUE TITLE
realSFS -sites

### ISSUE THREAD

**Issue by claudiuskerth** [ANGSD/angsd#81]
Hi ANGSD,

the documentation for `realSFS` on the ANGSD wiki says:

>You can also supply -sites directly to the realSFS subprogram for choosing a subset of sites. This could be useful if you are interested in the spectra for differenct functional categories.

I have tried this for the estimation of a 2D SFS:

>realSFS -P 12 -sites 0_boot.sites ../SAFs/ERY/ERY.unfolded.saf.idx ../SAFs/PAR/PAR.unfolded.saf.idx 2>/dev/null > SFS/0_boot.2dsfs &

I have provided a sites file that contains bootstrapped contigs, i. e. the sites for many contigs occur several times in the (sorted) sites file.

In the STDERR of the above `realSFS` command I can find the following:

>        -> Is in multi sfs, will now read data from chr:Contig_100060
        -> hello Im the master merge part of realSFS. and I'll now do a tripple bypass to find intersect 
        -> 1) Will set iter according to chooseChr and start and stop, and possibly using -sites

        -> Potential problem: The filereading has reached a chromsome: 'Contig_100060', which is not included in your '-sites' file.                                                       -> Please consider limiting your analysis to the chromsomes of interest 
        -> see 'http://www.popgen.dk/angsd/index.php/Sites' for more information
        -> Program will continue reading this chromosome... 
        -> Sites to keep[Contig_100060] from pop0:      67
        -> Sites to keep[Contig_100060] from pop1:      67
        -> [readdata] lastread:67 posi:7
        -> Comparing positions: 1 with 0 has:310
        -> Only read nSites: 310 will therefore prepare next chromosome (or exit)                                                                      
        -> Done reading data from chromosome will prepare next chromosome

When I compare the resulting 2dsfs with the 2dsfs previously estimated from all contigs, I see no difference. The spectra a practically identical. It looks as though `realSFS` had ignored the sites file.

I would be great if I would not have to reestimate the SAF's for each bootstrap replicate. I would also use the SAF file(s) for bootstrap replicates of contigs, but I cannot handle its format.

It would be really great if you could help me with this.

claudius

**Comment by ANGSD** [comment]
Hi Claudius could you try printing the regions that will be used in the analysis.
Something like
./realSFS print pop1.saf.idx pop2.saf.idx -sites yoursites
and validate that this looks correct.
The above commands prints to stdout so you can get the postions by piping into cut -f1-2

Best

**Comment by claudiuskerth** [comment]
Hi ANGSD,

many thanks for your quick reply!

I have run the following command line:
> realSFS print ../SAFs/ERY/ERY.unfolded.saf.idx ../SAFs/PAR/PAR.unfolded.saf.idx -sites 0_boot.sites 2>/dev/null | cut -f1-2 > extract

Number of sites in `extract`:
>wc -l extract
1130775 extract

Applying `-sites` had no effect on the number of extracted sites:
>realSFS print ../SAFs/ERY/ERY.unfolded.saf.idx ../SAFs/PAR/PAR.unfolded.saf.idx  2>/dev/null | wc -l
1130775

This file contains no replicate sites:
>sort extract | uniq | wc -l
1130775

Number of sites (including replicates) in `0_boot.sites`:
>wc - l 0_boot.sites
1733791 0_boot.sites

Number of unique sites in `0_boot.sites`:
>sort 0_boot.sites | uniq | wc -l
1100220

>sort -Vk1,1 -k2,2 0_boot.sites | uniq > uniq.boot.sites
>wc -l uniq.boot.sites
1100220

>angsd sites index uniq.boot.sites 2>/dev/null
>realSFS print ../SAFs/ERY/ERY.unfolded.saf.idx ../SAFs/PAR/PAR.unfolded.saf.idx -sites uniq.boot.sites 2>/dev/null | wc -l
1130775

Bottom line, the `-sites` option has no effect. All (overlapping) sites are printed.

hope that can give you a lead

claudius

**Comment by ANGSD** [comment]
Hello Claudius, i haven had time to look into this. But this is still on my
todo list.

Best

**Comment by ANGSD** [comment]
Dear Claudius sorry that this has been dragging on. I just tested it on some datasets. And it seems to work on my end. Did you use the latest version?

:~/angsd$ ./misc/realSFS print angsdput.saf.idx 2>/dev/null |wc -l
1667075
:~/angsd$ ./misc/realSFS print angsdput.saf.idx -sites sites.txt 2>/dev/null |wc -l
4730
:~/angsd$ ./misc/realSFS print angsdput.saf.idx angsdput.saf.idx -sites sites.txt 2>/dev/null |wc -l
4730
:~/angsd$ ./misc/realSFS print angsdput.saf.idx angsdput.saf.idx 2>/dev/null |wc -l
1667075

**Comment by claudiuskerth** [comment]
Hi Thorfinn,

I have just pulled the most recent version of `htslib` and `angsd`. I compiled `htslib` and installed it system wide. I then compiled `angsd`. Using the new `realSFS` and executing the commands in my previous post I see the same numbers, i. e. with a bootstrapped sites file, with a `uniq`'ed sites file or with no sites file, I always get the same number of sites from `realSFS print` (which is all overlapping sites). I have also created a smaller sites file with 5000 sorted sites, but still get all sites printed. In the STDERR I also still see this:

> -> Potential problem: The filereading has reached a chromsome: 'Contig_100028', which is not included in your '-sites' file.                                                                                -> Please consider limiting your analysis to the chromsomes of interest                                                                                                                                     -> see 'http://www.popgen.dk/angsd/index.php/Sites' for more information                                                                                                                                    -> Program will continue reading this chromosome...                                                                                                                                                         -> Sites to keep[Contig_100028] from pop0:      72                                                                                                                                                          -> Sites to keep[Contig_100028] from pop1:      72

I have attached the output of `make` from the compilation of `angsd`: [make.txt](https://github.com/ANGSD/angsd/files/1079065/make.txt). It emits a few warnings. 

I really hope you still have an idea how to make this work on my system. I am working on Ubuntu 14.04.2.

many thanks for your help!

claudius

**Comment by claudiuskerth** [comment]
Hi Thorfinn,

I had created the SAF files with an older version of `angsd`. Do you think it would be worth recreating them with the newest version of `angsd`?

claudius

**Comment by ANGSD** [comment]
Hi Claudius, there was indeed a but. Sorry for the time it took to fix. Thanks for reporting it.

Best

**Comment by claudiuskerth** [comment]
Hi Thorfinn,

thank you for fixing that bug. I can confirm that the `-sites` routine now works as described in the ANGSD wiki.

The following is a feature request:

The `-sites` routine currently cannot be used for custom bootstrapping the SFS. If presented with a sites file that contains multiple occurrences of the same site (as there should be in a bootstrap resample), `realSFS` only reads each site once (at least when called with `realSFS print`). I know there is a special routine in the `realSFS` programme called `-bootstrap`. I assume that this resamples over sites. However, resampling over tightly linked sites underestimates uncertainty and resampling over unlinked blocks of linked sites should provide better estimates of uncertainty. I think this would best be implemented via the `-sites` routine, so that users can provide custom bootstrapped sites files. All it needs is that `realSFS` accepts multiple occurrences of the same site for SFS estimation.

many thanks for your support!

claudius
